### PR TITLE
Add names for bundles and test they are consistent

### PR DESCRIPTION
**Pull request by stefpiatek**
Moved raw data to new structure with a separate file for the names, and added in some tests to ensure that we can detect inconsitencies. Bit of refactoring of raw data functions, expect that once we have a release serving off a package .rda file that we'll change the raw data to live outside of the package but seems fine for now as we still need to serve from the package.

- [ ] One question though, what domain does indices of deprivation go into?

### CODE REVIEW

**Comment by milanmlft** [comment]
> expect that once we have a release serving off a package .rda file that we'll change the raw data to live outside of the package but seems fine for now as we still need to serve from the package.

Agreed, having it run off `.rda` would mean all the data is immediately available and we don't need to re-read the csv files every time we wan to access a bundle. On the other hand, it might make loading the package slower as all the data would immediately be loaded in as well. If the data gets big enough that might become a performance issue. Maybe worth doing some benchmarking on that once we get to that point.

File: R/bundles.R [hunk 1]
```diff
@@ -13,31 +15,13 @@
...
 #' available_bundles("0.1")
 available_bundles <- function(version = "latest") {
-  raw_dir <- .get_raw_dir(version = version)
-  directories <- list.dirs(raw_dir, full.names = TRUE)
-  domain_directories <- directories[directories != raw_dir]
-
-  purrr::map_dfr(domain_directories, .build_concepts_from_directory) |>
-    mutate(version = version)
-}
+  raw_dir <- omopbundles:::get_raw_dir(version = version)
```
Comment (milanmlft) [inline R/bundles.R]:
```suggestion
  raw_dir <- get_raw_dir(version = version)
```

No need for `:::` as we're in the same package?

File: R/bundles.R [hunk 2]
```diff
@@ -13,31 +15,13 @@
...
-  concept_name <- NULL
-
-  dplyr::tibble(
-    id = concept_files,
-    concept_name = concept_files,
-    domain = basename(directory)
-  ) |>
-    dplyr::mutate(concept_name = sub("\\.csv$", "", concept_name)) |>
-    dplyr::mutate(concept_name = gsub("_", " ", concept_name))
+  purrr::map_dfr(bundle_name_paths, omopbundles:::parse_bundle_names) |>
```
Comment (milanmlft) [inline R/bundles.R]:
```suggestion
  purrr::map_dfr(bundle_name_paths, parse_bundle_names) |>
```

File: R/bundles.R [hunk 3]
```diff
@@ -52,12 +36,12 @@ available_bundles <- function(version = "latest") {
...
 #' # Usage with available_bundles, from a single row
-#' smoking_info <- available_bundles() |> dplyr::filter(concept_name == "smoking")
+#' smoking_info <- available_bundles() |> dplyr::filter(concept_name == "Smoking")
 #' concept_by_bundle(domain = smoking_info$domain, id = smoking_info$id, version = smoking_info$version)
 #' # Using if you know the details directly
-#' concept_by_bundle(domain = "observation", id = "smoking.csv")
+#' concept_by_bundle(domain = "observation", id = "smoking")
 concept_by_bundle <- function(domain, id, version = "latest") {
-  .get_raw_dir(version = version, domain, id) |>
+  omopbundles:::get_raw_dir(version = version, domain, "bundles", glue::glue("{id}.csv")) |>
```
Comment (milanmlft) [inline R/bundles.R]:
```suggestion
  get_raw_dir(version = version, domain, "bundles", glue::glue("{id}.csv")) |>
```

File: R/raw-data.R [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    dirname() |>
+    basename()
+
+  readr::read_csv(bundle_name_path, col_types = "cc") |>
+    mutate(domain = bundle_name)
+}
+
+
+
+#' @title Get concepts for a a single bundle row
```
Comment (milanmlft) [inline R/raw-data.R:50]:
```suggestion
#' @title Get concepts for a single bundle row
```

File: R/raw-data.R [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+
+# The package will store releases internally, and not in raw data
+# Until that happens, this will duplicate the bundle code
+
+#' @title Get available raw bundles
+#'
+#' @description If a bundle has multiple names, then the id will be duplicated across rows
+#'
+#' @return dataframe that contains a "concept_name" and a "domain" column for each available concept
```
Comment (milanmlft) [inline R/raw-data.R:15]:
```suggestion
#' @return dataframe that contains a "concept_name" and a "domain" column for each available concept
#' @keywords internal
```

Make explicit that this is a function for internal use only

File: R/raw-data.R [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+
+
+#' @title Get concepts for a a single bundle row
+#'
+#' @description Retrieves concept data for a specific bundle.
+#'
+#' @param domain The domain of the bundle.
+#' @param id The ID of the bundle.
+#' @return A data frame with the concept data.
```
Comment (milanmlft) [inline R/raw-data.R:56]:
```suggestion
#' @return A data frame with the concept data.
#' @keywords internal
```

File: tests/testthat/test-raw-data.R [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+  # get the domain
+  parts <- stringr::str_split(path, "/")[[1]]
+  domain <- parts[length(parts) - 2]
+
+  list(id = filename, domain = domain)
+}
+
+assert_bundle_has_name <- function(bundle) {
+  bundle_name_file <- omopbundles:::get_raw_dir(bundle$domain, "bundle_names.csv")
```
Comment (milanmlft) [inline tests/testthat/test-raw-data.R:20]:
```suggestion
  bundle_name_file <- get_raw_dir(bundle$domain, "bundle_names.csv")
```

File: tests/testthat/test-raw-data.R [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  bundle_names <- read_csv(bundle_name_file, show_col_types = FALSE)
+
+  expect_true(bundle$id %in% bundle_names$id,
+              glue::glue("{bundle$id} should at least one name in: {bundle$domain}/bundle_names.csv"))
+}
+
+test_that("All raw bundles have at least one name", {
+  raw_dir <- omopbundles:::get_raw_dir()
```
Comment (milanmlft) [inline tests/testthat/test-raw-data.R:28]:
```suggestion
  raw_dir <-get_raw_dir()
```

File: tests/testthat/test-raw-data.R [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+  concept_files <- Sys.glob(file.path(raw_dir, "*", "bundles", "*.csv"))
+  bundle_ids <- purrr::map(concept_files, extract_bundle_details)
+  purrr::walk(bundle_ids, assert_bundle_has_name)
+})
+
+
+
+test_that("All raw bundle names map to a bundle file that has at least one concept", {
+  bundles <- omopbundles:::raw_bundles()
```
Comment (milanmlft) [inline tests/testthat/test-raw-data.R:38]:
```suggestion
  bundles <- raw_bundles()
```

File: tests/testthat/test-raw-data.R [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+  # Ensure bundles dataframe is not empty
+  expect_true(nrow(bundles) > 0)
+
+  apply(bundles, 1, function(bundle) {
+    concepts <- omopbundles:::raw_concept_by_bundle(bundle["domain"], bundle["id"])
```
Comment (milanmlft) [inline tests/testthat/test-raw-data.R:44]:
```suggestion
    concepts <- raw_concept_by_bundle(bundle["domain"], bundle["id"])
```

File: tests/testthat/test-raw-data.R [hunk 11]
```diff
@@ -0,0 +1,50 @@
+library(dplyr)
+library(omopbundles)
+library(testthat)
+library(stringr)
+library(purrr)
+
+
+
```
Comment (milanmlft) [inline tests/testthat/test-raw-data.R:8]:
```suggestion
```

Shouldn't have `library()` calls in tests

File: tests/testthat/test-bundles.R [hunk 12]
```diff
@@ -6,31 +6,31 @@ library(testthat)
 test_that("available_bundles isn't empty and have correct columns", {
   result <- omopbundles::available_bundles()
   expect_true(nrow(result) > 0, info = "The dataframe should not be empty")
-  hiv_ab <- filter(result, concept_name == "antibodies to hiv")
+  hiv_ab <- filter(result, concept_name == "Antibodies to HIV")
```
Comment (milanmlft) [inline tests/testthat/test-bundles.R:9]:
```suggestion
  hiv_ab <- dplyr::filter(result, concept_name == "Antibodies to HIV")
```

Make sure we use `dplyr`'s filter and not `stats::filter`

File: tests/testthat/test-bundles.R [hunk 13]
```diff
@@ -6,31 +6,31 @@ library(testthat)
...
-    dplyr::filter(concept_name == "smoking")
+    dplyr::filter(concept_name == "Smoking")
 
   expect_true(nrow(result) == 1, info = "Smoking should only exist as a single row")
   expect_equal(result$domain, "observation")
 })
 
 test_that("Concept by bundle works with character values", {
-  smoking_concepts <- omopbundles::concept_by_bundle(domain = "observation", id = "smoking.csv")
+  smoking_concepts <- omopbundles::concept_by_bundle(domain = "observation", id = "smoking")
```
Comment (milanmlft) [inline tests/testthat/test-bundles.R:25]:
```suggestion
  smoking_concepts <- concept_by_bundle(domain = "observation", id = "smoking")
```

File: R/raw-data.R [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+
+get_raw_dir <- function(..., version = "latest") {
+  if (version != "latest") warning("Versioning not yet implemented, using version = 'latest'")
+
+  file_path <- system.file("data-raw", ..., package = "omopbundles")
+
+  if (!file.exists(file_path)) {
+    path <- paste(..., sep = "/")
```
Comment (milanmlft) [inline R/raw-data.R:31]:
```suggestion
    path <- file.path(...)
```

To make it OS-agnostic (i.e. deal with Windows...) better to use `file.path()`, which will put the appropriate separator based on the OS.

Comment (milanmlft) [reply R/raw-data.R:31]: Ah wait, never mind, didn't realise this was just for the error messager

**Review by milanmlft — APPROVED** [review]
LGTM! 🚀 Just some suggestions regarding `library()` and `::`/`:::` calls
Just wondering how the `bundle_names.csv` files get generated? Might be worth adding a script to reproduce these files to the package. A common place for this could be `data-raw/bundle_names.R`

File: tests/testthat/test-raw-data.R (hunk 7 above)
Comment (stefpiatek) [reply tests/testthat/test-raw-data.R:20]: Will this work, as its not an exported function?

Comment (milanmlft) [reply tests/testthat/test-raw-data.R:12]: Yep, the `testthat` suite loads the entire package, just like you would do with `devtools::load_all()`
I think as a general rule, `:::` should be avoided as much as possible. I see it more as an interactive tool to inspect non-exported code from other packages if I need to debug something. Pretty sure CRAN doesn't even allow you to have any `:::` at all in your package.

Comment (stefpiatek) [reply tests/testthat/test-raw-data.R:12]: Ah interesting, I must be misremembering testing using internal functions, okay great

**Comment by stefpiatek** [comment]
> Just wondering how the `bundle_names.csv` files get generated? Might be worth adding a script to reproduce these files to the package. A common place for this could be `data-raw/bundle_names.R`

I think the names should be manually curated, or exported from the source system

**Pull request merged**